apiservice/apiservice.go

```diff
@@ -0,0 +1,240 @@
+package apiservice
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow/protobuf/go/flow/access"
+
+	"github.com/onflow/flow-go/engine/access/rpc/backend"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/utils/grpcutils"
+)
+
+func NewFlowAPIService(accessNodeAddressAndPort flow.IdentityList, timeout time.Duration) (*FlowAPIService, error) {
+	accessClients := make([]access.AccessAPIClient, accessNodeAddressAndPort.Count())
+	for i, identity := range accessNodeAddressAndPort {
+		if identity.NetworkPubKey == nil {
+			clientRPCConnection, err := grpc.Dial(
+				identity.Address,
+				grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(grpcutils.DefaultMaxMsgSize)),
+				grpc.WithInsecure(), //nolint:staticcheck
+				backend.WithClientUnaryInterceptor(timeout))
+			if err != nil {
+				return nil, err
+			}
+
+			accessClients[i] = access.NewAccessAPIClient(clientRPCConnection)
+		} else {
+			tlsConfig, err := grpcutils.DefaultClientTLSConfig(identity.NetworkPubKey)
+			if err != nil {
+				return nil, err
+			}
+
+			clientRPCConnection, err := grpc.Dial(
+				identity.Address,
+				grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(grpcutils.DefaultMaxMsgSize)),
+				grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)),
+				backend.WithClientUnaryInterceptor(timeout))
+			if err != nil {
+				return nil, err
+			}
+
+			accessClients[i] = access.NewAccessAPIClient(clientRPCConnection)
+		}
+	}
+
+	ret := &FlowAPIService{}
+	ret.upstream = accessClients
+	ret.roundRobin = 0
+	return ret, nil
+}
+
+type FlowAPIService struct {
+	access.AccessAPIServer
+	lock       sync.Mutex
+	roundRobin int
+	upstream   []access.AccessAPIClient
+}
+
+func (h *FlowAPIService) SetLocalAPI(local access.AccessAPIServer) {
+	h.AccessAPIServer = local
+}
+
+func (h *FlowAPIService) client() (access.AccessAPIClient, error) {
+	if h.upstream == nil || len(h.upstream) == 0 {
+		return nil, status.Errorf(codes.Unimplemented, "method not implemented")
+	}
+
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	h.roundRobin++
+	h.roundRobin = h.roundRobin % len(h.upstream)
+	ret := h.upstream[h.roundRobin]
```

Comment:
I'm curious how much we implement following gRPC abstractions: 
* https://pkg.go.dev/google.golang.org/grpc/resolver
* https://pkg.go.dev/google.golang.org/grpc/resolver/manual
* https://pkg.go.dev/google.golang.org/grpc/balancer
* https://pkg.go.dev/google.golang.org/grpc/balancer/roundrobin

Current implementation will work fine at first, but it would eventually show some problems:
* (round robin balancer) if one backend slows down, we would still be queueing more and more traffic to it, creating a negative feedback loop.
* (resolver) if access node dies (or gRPC client we've created above gets in a weird state) we would fail/timeout 1/nth of requests.
* (resolver) if a new node appears (or a dead node recovers after the startup) we won't be able to put it into the LB pool until the restart.

Reply:
Yes, it makes sense. This is a matter of negotiation. Neither of the issues that are the basis of this PR explained such behavior: [2063][2064][2066][2075][State Sync]. We were asked to be conservative adding new logic.

I am happy to update the PR or add a new one that implements gRPC load balancing, if negotiated with @AliSMAmin.

Personally, I would opt for having a simple 1-N round robin behavior like we have here with a hard panic() on any errors.
This will allow any hardware or Kubernetes load balancers or autoscalers deal with excess traffic, in an even more common and standard fashion than gRPC. It helps with debugging time, standardizing scaling logic. All in all, the goal is reducing ops costs.
In terms of reliability, if the best approach is used, such as retries are implemented in the gRPC client endpoint, it does not really matter, which layer faults (hardware load balancing, gRPC load balancing, Kubernetes, DNS based solution), as long as it is self-healing.

It is up to your consideration. Please let us know, what you decide.

Reply:
Thanks for the call out Alexey. This is useful although since we're currently blocked by this task and we need to get things moving we'll revisit this at a later point. Tracking it in https://github.com/onflow/flow-go/issues/2606

Reply:
I addressed the basic retry in: https://github.com/GetElastech/flow-go/commit/368497cdbc5574eb5af0e10d4d54561222e8aa2a